src/lua/LuaFileSystem.cpp

```diff
@@ -7,15 +7,112 @@
 #include "LuaObject.h"
 #include "Pi.h"
 
-/*
- * Interface: FileSystem
- *
- * A global table that provides access to the filesystem.
- *
- * This interface is protected. It can only be used by scripts in Pioneer's
- * data directory. Mods and scripts in the user directory that try to use it
- * will get a Lua error.
- */
+#include "core/StringUtils.h"
+
+class ParsedFilePath
+{
+public:
+	ParsedFilePath(lua_State *l, const char *luaPath);
+
+	// Probably not needed as if it isn't valid we'll already fail.
+	bool IsValid() const { return m_fs != nullptr; }
+
+	// The FileSourceFS structure allows us to write, so to grab one
+	// we validate that we can and error if we can't and return null
+	// otherewise we return the FileSourceFS.
+	FileSystem::FileSourceFS* ValidateWritePermission(lua_State *l) const;
+	FileSystem::FileSource&	GetFileSource() const { return *m_fs; }
+	const std::string_view& GetRelativePath() const { return m_fs_path; }
+
+	std::string CalculateAbsolutePath(lua_State* l) const;
+private:
+	std::string_view          m_luaPath;
+	std::string_view          m_scheme;
+	std::string_view          m_fs_path;
+	FileSystem::FileSource    *m_fs;
+	FileSystem::FileSourceFS  *m_writeableFS;
+};
+
+ParsedFilePath::ParsedFilePath(lua_State *l, const char *luaPath)
+	: m_luaPath(luaPath)
+{
+	if (starts_with_ci(m_luaPath, "user://"))
+	{
+		m_fs = &FileSystem::userFiles;
+		m_writeableFS = &FileSystem::userFiles;
+		m_scheme = m_luaPath.substr(0, 7);// strlen("user://"));
+	}
+	else if (starts_with_ci(m_luaPath, "data://"))
+	{
+		m_fs = &FileSystem::gameDataFiles;
+		m_writeableFS = nullptr;
+		m_scheme = m_luaPath.substr(0, 7);// strlen("data://"));
+	}
+	else
+	{
+		luaL_error(l, "'%s' does not have a valid scheme, must be user:// or data://", luaPath);
+		m_writeableFS = nullptr;
+		m_fs = nullptr;
+		return;
+	}
+	m_fs_path = m_luaPath;
+	m_fs_path.remove_prefix(m_scheme.length());
+}
+
+FileSystem::FileSourceFS* ParsedFilePath::ValidateWritePermission(lua_State *l) const
+{
+	if (m_writeableFS != nullptr)
+	{
+		return m_writeableFS;
+	}
+	luaL_error(l, "'%s' does not support file operations that modify it, only things in the user folder do", m_luaPath.data());
```

Comment:
By the by, `luaL_error` is a termination point for the function and should be treated as though it has `noreturn` semantics. The following `return` statements will never be reached, and there's a potential for memory leaks if you allocate something on the heap in the function that isn't automatically destroyed before the call to `luaL_error`.

For this reason, I'd highly recommend only calling `luaL_error` directly in a lua CFunction, using the idiom `return luaL_error(...);`.